include/hal/dts-hal.sh

```diff
@@ -0,0 +1,197 @@
+#    for dmidecode -s system-vendor it will be "-s system-vendor".
+#
+# This function is a bridge between common DTS logic and hardware-specific DTS
+# logic or functions. There is two paths a call to this function can be
+# redirected to: real HAL for running on real platform and Tests HAL for testing
+# on QEMU (depends on whether the var. DTS_TESTING is set or not).
+#
+# The real HAL are the real tools e.g. cbfstool, etc.. The testing HAL are the
+# mocking functions. There are several types of mocking functions, with every
+# type having a specific name syntax:
+#
+# FUNCTIONNAME_mock(){...}: mocking functions specific for every platform, those
+# are stored in $DTS_MOCK_PLATFORM file which is sourced at the beginning of
+# this file.
+# TOOLNAME_FUNCTIONNAME_mock(){...}: mocking functions common for all platforms
+# but specific for some tool, those are stored in $DTS_MOCK_COMMON file, which
+# is being sourced at the beginning of this file.
+# TOOLNAME_common_mock(){...}: standard mocking functions for every tool that
+# are common for all platforms, those are stored in $DTS_MOCK_COMMON file, which
+# is being sourced at the beginning of this file.
+# common_mock(){...}: common mocking function, in case we need to use mocking
+# function for a tool but we do not care about its output.
+#
+# This tool wrapper should only be used with tools which communicate with
+# hardware or firmware (read or write, etc.).
+#
+# TODO: this wrapper deals with arguments as well as with stdout, stderr, and $?
+# redirection, but it does not read and redirect stdin (this is not used in any
+# mocking functions or tools right now).
+  # Gets toolname, e.g. poweroff, dmidecode. etc.:
+  local _tool="$1"
+  # Gets mocking function name:
+  local _mock_func="$2"
+  # It checks if _mock_func contains smth with _mock at the end, if not -
+  # mocking function is not provided and some common mocking func. will be used
+  # instead:
+  if ! echo "$_mock_func" | grep "_mock" &> /dev/null; then
+    unset _mock_func
+    shift 1
+  else
+    shift 2
+  fi
+  # Other arguments for this function are the arguments which are sent to a tool
+  # e.g. -s system-vendor for dmidecode, etc.:
+  local _arguments=( "$@" )
+
+  if [ -n "$DTS_TESTING" ]; then
+    # This is the order of calling mocking functions:
+    # 1) FUNCTIONNAME_mock;
+    # 2) TOOLNAME_FUNCTIONNAME_mock;
+    # 3) TOOLNAME_common_mock;
+    # 4) common_mock - last resort.
+    if [ -n "$_mock_func" ] && type $_mock_func &> /dev/null; then
+      $_mock_func "${_arguments[@]}" 1>&1 2>&2
+    elif type ${_tool}_${_mock_func} &> /dev/null; then
+      ${_tool}_${_mock_func} "${_arguments[@]}" 1>&1 2>&2
+    elif type ${_tool}_common_mock &> /dev/null; then
+      ${_tool}_common_mock "${_arguments[@]}" 1>&1 2>&2
+    else
+      common_mock $_tool 1>&1 2>&2
+    fi
+
+    return $?
+  fi
+
+  # If not testing - call tool with the arguments instead:
+  $_tool "${_arguments[@]}" 1>&1 2>&2
```

Comment:
I wanted to show explicitly what goes where here, because this function may not be so clear at first glance, therefore I have left these redirections.

Reply:
It makes it even less clear. I never seen `1>&1` and from what I can understand it does nothing.
On first read I thought you wanted to redirect `stderr` to `stdout` and maybe `stdout` to somewhere else but made mistake.

Reply:
Addressed [here](https://github.com/Dasharo/dts-scripts/compare/e748126a96325d132585966533a906be7dcd2597..a29856f198488767178cbe8e15853f96def1fb8b)